[SPARK-9112] [ML] Implement Stats for LogisticRegression

MechCoder opened the pull request:
I have added support for stats in LogisticRegression. The API is similar to that of LinearRegression with LogisticRegressionTrainingSummary and LogisticRegressionSummary

I have some queries and asked them inline.

MechCoder committed:
[SPARK-9112] [ML] Implement Stats for LogisticRegression

MechCoder commented:
@feynmanliang

AmplabJenkins commented:
Merged build triggered.

MechCoder commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,60 @@ private[classification] class MultiClassSummarizer extends Serializable {
+   * which is an RDD of (false positive rate, true positive rate)
+   * with (0.0, 0.0) prepended and (1.0, 1.0) appended to it.
+   */
+  val roc: RDD[(Double, Double)] = metrics.roc()
```
Should this be returned as a dataframe?

feynmanliang replied:
Does this even need to be a distributed data structure?

MechCoder replied:
It seems that the size of the ROC curve and all other metrics is equal to the size of the data. (i.e it chooses every possible score as a threshold)

MechCoder replied:
This is controlled by the numBins parameter (that I did not see). Any idea how to make this accessible to the user?

MechCoder commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,60 @@ private[classification] class MultiClassSummarizer extends Serializable {
+
+@Experimental
+class LogisticRegressionSummary private[classification] (
+  @transient val predictions: DataFrame,
```
Is this transient because we do not want to store the predictions?

MechCoder replied:
Is this transient because we do not want to store the predictions?

feynmanliang replied:
Yes.

Since models will be serialized to disk (model save/load) and sent to executors (during `model.predict` closure capture) we wanted to avoid movement of large and expensive data.

Small/cheap summary statistics (e.g. scalars precision, recall, auroc) should be available in the `*Model` instance but large ones (e.g. residuals in LinearRegressionSummary) should only be available in the driver and not be serialized (i.e. transient).

MechCoder commented on mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala:
```diff
@@ -701,4 +701,19 @@ class LogisticRegressionSuite extends SparkFunSuite with MLlibTestSparkContext {
+
+    val sameSummary = model.evaluate(dataset)
+    assert(summary.areaUnderROC === sameSummary.areaUnderROC)
+
```
TODO: Add more tests, after the first pass has been done.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37832 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37832/consoleFull) for   PR 7538 at commit [`70a0fc4`](https://github.com/apache/spark/commit/70a0fc42e8aad7ca78cd97d1a77170a7d422799d).

SparkQA commented:
[Test build #37832 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37832/console) for   PR 7538 at commit [`70a0fc4`](https://github.com/apache/spark/commit/70a0fc42e8aad7ca78cd97d1a77170a7d422799d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -286,6 +294,40 @@ class LogisticRegressionModel private[ml] (
+  private var trainingSummary: Option[LogisticRegressionTrainingSummary] = None
+
+  /**
+   * Gets summary (e.g. residuals, mse, r-squared ) of model on training set. An exception is
```
Update doc

MechCoder replied:
Haha, this is embarassing :P

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -286,6 +294,40 @@ class LogisticRegressionModel private[ml] (
+    case Some(summ) => summ
+    case None =>
+      throw new SparkException(
+        "No training summary available for this LinearRegressionModel",
```
Update name

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -286,6 +294,40 @@ class LogisticRegressionModel private[ml] (
+   * @param dataset Test dataset to evaluate model on.
+   */
+  // TODO: decide on a good name before exposing to public API
+  def evaluate(dataset: DataFrame): LogisticRegressionSummary = {
```
Make this private until we stabilize the API for summaries

MechCoder replied:
sure. I will change it just before merge. I kept it public so that I can access it in the shell.

feynmanliang replied:
Can you make this private in the PR so I can give LGTM?

feynmanliang commented:
Made a first pass.

Maybe it might make sense to make a `HasSummary` trait with the `hasSummary` and `setSummary` logic for `*Model`s to mix-in? Then we can introduce a `ModelSummary` trait implemented by all the summary objects.

feynmanliang commented:
Forgot to add, would be nice to include `@transient confusion : RDD[(threshold: Double, confusionMatrix: BinaryConfusionMatrix)]` in `LogisticRegressionSummary` since it's a good diagnostic tool despite already being summarized by precision/recall/etc (although may be hard to get this into a DataFrame since BinaryConfusionMatrix doesn't have a good UDT, @jkbradley ?) ... We do that in `LinearRegressionSummary` with residuals

MechCoder commented:
Thanks a lot for your kind reviews :)

> Maybe it might make sense to make a HasSummary trait with the hasSummary and setSummary

Yes, indeed. Where should such a trait go? Should we have a ml/summary ? Would it better to refactor this in a different PR or this one?

Also It might help to make a `RegressionSummary` or a `ClassificationSummary` after this has been done because most of the regression metrics and classification metrics are common to different ML algorithms.

MechCoder commented:
Btw, I assume it had been decided not to add hinge loss, log loss etc ?

MechCoder commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -252,7 +254,13 @@ class LogisticRegression(override val uid: String)
     if (handlePersistence) instances.unpersist()
 
-    copyValues(new LogisticRegressionModel(uid, weights, intercept))
+    val model = copyValues(new LogisticRegressionModel(uid, weights, intercept))
```
Can you please explain why this `copyValues` is necessary? and I'm unable to understand how $(probabilityCol) gives a string because when I do this.

    val model = lr.fit(dataset)
    $(lr.probabilityCol)

I get 

    error: not found: value $
    $(probabilityCol)

feynmanliang replied:
`$` is defined in `Params`, which `LogisticRegression` mixes in via `LogisticRegressionParams`.

See https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/param/params.scala#L463

MechCoder replied:
Ah, I see thanks !

MechCoder committed:
DataFrame support for metrics

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37947 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37947/consoleFull) for   PR 7538 at commit [`fbed861`](https://github.com/apache/spark/commit/fbed861eace8d0fe780ef326ee9b9164982a1ac5).

SparkQA commented:
[Test build #37947 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37947/console) for   PR 7538 at commit [`fbed861`](https://github.com/apache/spark/commit/fbed861eace8d0fe780ef326ee9b9164982a1ac5).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

MechCoder committed:
Added tests

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37952 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37952/consoleFull) for   PR 7538 at commit [`80d9954`](https://github.com/apache/spark/commit/80d995424d06e353b907b9fde1c99bf81f9538a9).

MechCoder commented:
I've addressed your comments about the dataframe storage.

SparkQA commented:
[Test build #37952 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37952/console) for   PR 7538 at commit [`80d9954`](https://github.com/apache/spark/commit/80d995424d06e353b907b9fde1c99bf81f9538a9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

MechCoder commented:
@feynmanliang any news on this? thanks.

feynmanliang commented:
@MechCoder sorry for the delays! We are having a hackathon at my work; I will review when I am in the office on monday.

MechCoder commented:
okay, thanks :)

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,99 @@ private[classification] class MultiClassSummarizer extends Serializable {
+   * Every possible probability obtained in transforming the dataset are used
+   * as thresholds used in calculating the FPR and TPR.
+   */
+  @transient val roc: DataFrame = {
```
This should be a `def`; we only want to lazily evaluate `roc` if the user asks for it (same thing is going on in `BinaryClassificationMetrics`). Ditto for others

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,99 @@ private[classification] class MultiClassSummarizer extends Serializable {
+  @transient val metrics = new BinaryClassificationMetrics(
+    predictions.select(probabilityCol, labelCol).map {
+      case Row(score: Vector, label: Double) => (score(1), label)
+    }
```
We should provide `numBins`  rather than rely on the `0` default which creates one threshold per data-point.

MechCoder replied:
Indeed. But how do we expose this to the user. Or do we use some sort of heuristic?

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,99 @@ private[classification] class MultiClassSummarizer extends Serializable {
+   */
+  @transient val roc: DataFrame = {
+    val distributedRoc = metrics.roc()
+    val sqlContext = SQLContext.getOrCreate(distributedRoc.sparkContext)
```
Let's just use the `predictions` sqlContext directly

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,99 @@ private[classification] class MultiClassSummarizer extends Serializable {
+  @transient val roc: DataFrame = {
+    val distributedRoc = metrics.roc()
+    val sqlContext = SQLContext.getOrCreate(distributedRoc.sparkContext)
+    import sqlContext.implicits._
```
Wrapping this stuff into a `DataFrame` is repeated quite a bit, maybe introduce a helper method?

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,99 @@ private[classification] class MultiClassSummarizer extends Serializable {
   }
 }
 
+@Experimental
```
scaladocs

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,99 @@ private[classification] class MultiClassSummarizer extends Serializable {
+}
+
+@Experimental
+class LogisticRegressionSummary private[classification] (
```
scaladocs

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,99 @@ private[classification] class MultiClassSummarizer extends Serializable {
+
+  /** Returns a BinaryClassificationMetrics object.
+  */
+  @transient val metrics = new BinaryClassificationMetrics(
```
private

feynmanliang commented on mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala:
```diff
@@ -701,4 +701,36 @@ class LogisticRegressionSuite extends SparkFunSuite with MLlibTestSparkContext {
     assert(model1.intercept ~== interceptR relTol 1E-5)
     assert(model1.weights ~= weightsR absTol 1E-6)
```
Not your fault, but this should be `~==`

MechCoder committed:
remove unnecessary dataframe stuff and add docs

MechCoder commented:
@feynmanliang thanks for your comments. I've addressed most of them (except for the bins_

Is it agreed that the refactoring can be done in another PR?

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38697 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38697/consoleFull) for   PR 7538 at commit [`640376a`](https://github.com/apache/spark/commit/640376a2c7a6b033f71b954cf44f0428cd6cfb0b).

SparkQA commented:
[Test build #38697 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38697/console) for   PR 7538 at commit [`640376a`](https://github.com/apache/spark/commit/640376a2c7a6b033f71b954cf44f0428cd6cfb0b).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

feynmanliang commented:
Spark 1.5 is shipping soon so if you want it to be included I would get all the changes into this PR.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,98 @@ private[classification] class MultiClassSummarizer extends Serializable {
+    metrics.fMeasureByThreshold().toDF("threshold", "F-Measure")
+  }
+
+  /** Returns a dataframe with two fields (threshold, precision) curve.
```
Use JavaDoc style, see [style guide](https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide#SparkCodeStyleGuide-Codedocumentationstyle). Ditto for others

feynmanliang commented on mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala:
```diff
@@ -699,6 +699,40 @@ class LogisticRegressionSuite extends SparkFunSuite with MLlibTestSparkContext {
+      model.summary
+        .objectiveHistory
+        .sliding(2)
+        .forall(x => x(0) >= x(1)))
```
nit: state what is being tested ("lossHistory is monotonically decreasing") in comment

feynmanliang commented:
LGTM after the minor changes. @jkbradley for final pass.

MechCoder commented:
Oh, by refactoring I had meant the `ModelSummary` trait which can be inherited by these. Also the question about exposing the `numBins parameter` . It might be good to have a setBins parameter in the future so that the user can reuse the instance of `BinaryClassificationMetrics` and we can wrap it around here.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,98 @@ private[classification] class MultiClassSummarizer extends Serializable {
+  @transient private val metrics = new BinaryClassificationMetrics(
+    predictions.select(probabilityCol, labelCol).map {
+      case Row(score: Vector, label: Double) => (score(1), label)
+    }
```
For now, please default numBins to something sensible like 100

feynmanliang commented:
Oh, yeah the trait and `numBins` setter can be done in follow up work. For now, though, we should set the numBins explicitly (see in-line comment). Can you create the follow up JIRAs? Thanks.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

MechCoder commented:
retest this please

MechCoder commented:
@jkbradley can this make 1.5? thanks !

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #161 has started](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/161/consoleFull) for   PR 7538 at commit [`e4735a4`](https://github.com/apache/spark/commit/e4735a4ac490c51d1a2a5da02569b8b16eb560cc).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #161 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/161/console) for   PR 7538 at commit [`e4735a4`](https://github.com/apache/spark/commit/e4735a4ac490c51d1a2a5da02569b8b16eb560cc).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #39035 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39035/consoleFull) for   PR 7538 at commit [`e4735a4`](https://github.com/apache/spark/commit/e4735a4ac490c51d1a2a5da02569b8b16eb560cc).

SparkQA commented:
[Test build #39035 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39035/console) for   PR 7538 at commit [`e4735a4`](https://github.com/apache/spark/commit/e4735a4ac490c51d1a2a5da02569b8b16eb560cc).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

MechCoder committed:
minor

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

MechCoder commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #187 has started](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/187/consoleFull) for   PR 7538 at commit [`40ad8ef`](https://github.com/apache/spark/commit/40ad8ef682b9971d44ed30411d668c8ce10c4d25).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39366 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39366/consoleFull) for   PR 7538 at commit [`40ad8ef`](https://github.com/apache/spark/commit/40ad8ef682b9971d44ed30411d668c8ce10c4d25).

SparkQA commented:
[Test build #39366 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39366/console) for   PR 7538 at commit [`40ad8ef`](https://github.com/apache/spark/commit/40ad8ef682b9971d44ed30411d668c8ce10c4d25).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #187 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/187/console) for   PR 7538 at commit [`40ad8ef`](https://github.com/apache/spark/commit/40ad8ef682b9971d44ed30411d668c8ce10c4d25).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

jkbradley commented:
@MechCoder Just getting to this now, sorry!  It looks quite clean.  My main question is how we should handle multiclass LogisticRegression.  Any ideas?  I could imagine 2 main options:
* (A) Include the union of MulticlassMetrics and BinaryClassificationMetrics in the summary.  For multiclass data, have empty/bad values (or exceptions) for binary classification metrics which do not make sense for multiclass.
* (B) Create a LogisticRegressionSummary which has multiclass metrics.  Also create a BinaryLogisticRegressionSummary which extends LogisticRegressionSummary and provides binary metrics.

I'd prefer (B).  That could be done all in this PR, or it could be split into 2 (with the first only adding LogisticRegressionSummary, and the second adding BinaryLogisticRegressionSummary).

CC:  @feynmanliang

feynmanliang commented:
Is it still called logistic regression when there's multiple classes (I've
always called the multi class version softmax)?

In any case, I think this pr is fine since the logistic regression pipeline
stage being modified is a binary classifier. In the future, we should
definitely generalize to something like classifierSummary and
regressionSummary

On Sat, Aug 1, 2015, 5:59 PM jkbradley <notifications@github.com> wrote:

> @MechCoder <https://github.com/MechCoder> Just getting to this now,
> sorry! It looks quite clean. My main question is how we should handle
> multiclass LogisticRegression. Any ideas? I could imagine 2 main options:
>
>    - (A) Include the union of MulticlassMetrics and
>    BinaryClassificationMetrics in the summary. For multiclass data, have
>    empty/bad values (or exceptions) for binary classification metrics which do
>    not make sense for multiclass.
>    - (B) Create a LogisticRegressionSummary which has multiclass metrics.
>    Also create a BinaryLogisticRegressionSummary which extends
>    LogisticRegressionSummary and provides binary metrics.
>
> I'd prefer (B). That could be done all in this PR, or it could be split
> into 2 (with the first only adding LogisticRegressionSummary, and the
> second adding BinaryLogisticRegressionSummary).
>
> CC: @feynmanliang <https://github.com/feynmanliang>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/7538#issuecomment-126970026>.
>

MechCoder commented:
@jkbradley Doesn't `LogisticRegression` just handle binary labels for now? I think we can just leave handling of multiclass as and when it is supported.

But for now, I think we should have a `ml/summary` and refactor it into `RegressionSummary`, `ClassificationSummary` and reuse it as much as possible as I cannot think of any specific metrics that would be suitable only for `LogisticRegression` 

I can do that after this as merged.

jkbradley commented:
@feynmanliang "Logistic regression" refers to both binary & multiclass.

LogisticRegression handles multiclass, so we'll need to make it work for multiclass from the start.

MechCoder commented:
Oh, I got mislead by this (https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala#L49) then

MechCoder commented:
Actually I'm not able to find a single test that checks for multiclass input. It says `generateMultinomialLogisticInput` here (https://github.com/apache/spark/blob/master/mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala#L59) but it seems that it is actually a binary dataset.

Also https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala#L287 here numClasses is set explicitly to 2.

Surely I'm missing something?

jkbradley commented:
Uh, yeah, you're right; it is only for binary.  Sorry!

MechCoder commented:
So I can haz merge / reviews? It is all right, if you are busy with 1.5 though.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,103 @@ private[classification] class MultiClassSummarizer extends Serializable {
+
+}
+
+@Experimental
```
This annotation should go after Scala doc, right before the class header.  (same for the one above too)

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,103 @@ private[classification] class MultiClassSummarizer extends Serializable {
+ *                       each sample.
+ * @param labelCol field in "predictions" which gives the true label of each sample.
+ */
+class LogisticRegressionSummary private[classification] (
```
Before merging, I still think we need to create the abstraction to handle both binary and multiclass classification summaries.  If we commit this PR now, then we will need to break the API in the future to support multiclass classification.  Instead, we should prepare for it now by providing the abstraction + the binary summary in this PR.  (It's fine not to have the multiclass version in this PR.)

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,103 @@ private[classification] class MultiClassSummarizer extends Serializable {
+  private val sqlContext = predictions.sqlContext
+  import sqlContext.implicits._
+
+  /** Returns a BinaryClassificationMetrics object.
```
fix scala doc style

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,103 @@ private[classification] class MultiClassSummarizer extends Serializable {
+   * which is an Dataframe having two fields (false positive rate, true positive rate)
+   * with (0.0, 0.0) prepended and (1.0, 1.0) appended to it.
+   * Every possible probability obtained in transforming the dataset are used
+   * as thresholds used in calculating the FPR and TPR.
```
It'd be nice to include the Wikipedia link for ROC

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,103 @@ private[classification] class MultiClassSummarizer extends Serializable {
+   * Every possible probability obtained in transforming the dataset are used
+   * as thresholds used in calculating the FPR and TPR.
+   */
+  def roc(): DataFrame = metrics.roc().toDF("FalsePositiveRate", "TruePositiveRate")
```
How about using names "FPR" and "TPR" instead?

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,103 @@ private[classification] class MultiClassSummarizer extends Serializable {
+
+  /**
+   * Returns the precision-recall curve, which is an Dataframe containing
+   * two fields (recall, precision) NOT (precision, recall), with (0.0, 1.0) prepended to it.
```
No need to specify the order of the fields since users should select by name, not index.

MechCoder replied:
right.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -407,6 +449,103 @@ private[classification] class MultiClassSummarizer extends Serializable {
+  /**
+   * Returns the precision-recall curve, which is an Dataframe containing
+   * two fields (recall, precision) NOT (precision, recall), with (0.0, 1.0) prepended to it.
+   * Every possible probability obtained in transforming the dataset are used
```
I would remove this sentence, and other copies of it below.  I don't think it's true since metrics uses binning, and I think modifying it would require more explanation.

MechCoder replied:
Yes, but this was true before I `setBins` to 100 right? I'll remove it anyway. We can allow user to setBins a future PR. How does that sound?

jkbradley replied:
True, it would be true without binning.  I agree we can probably allow setting bins in the future.

jkbradley commented:
@MechCoder It looks good, except for the issue with future multiclass support.

MechCoder committed:
Add abstraction to handle Multiclass Metrics

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

MechCoder commented:
@jkbradley thanks for your comments. I think I have addressed them in the last commit (including the multiclass support). 

Is the idea to set different summaries according to the data given?

AmplabJenkins commented:
Merged build finished. Test FAILed.

MechCoder commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #194 has started](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/194/consoleFull) for   PR 7538 at commit [`9586125`](https://github.com/apache/spark/commit/958612537d8ab41b72a976a371dcf462b6d733ba).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39581 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39581/consoleFull) for   PR 7538 at commit [`9586125`](https://github.com/apache/spark/commit/958612537d8ab41b72a976a371dcf462b6d733ba).

SparkQA commented:
[Test build #194 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/194/console) for   PR 7538 at commit [`9586125`](https://github.com/apache/spark/commit/958612537d8ab41b72a976a371dcf462b6d733ba).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #39581 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39581/console) for   PR 7538 at commit [`9586125`](https://github.com/apache/spark/commit/958612537d8ab41b72a976a371dcf462b6d733ba).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

jkbradley commented:
@MechCoder Thanks for the updates!  I'll take a look.  The issue with multiclass is that, when we add multiclass support in the future, the current (binary) metrics will not all be applicable to multiclass problems.  That means we'd need to return null or some bad value for certain metrics.  I'd prefer to use a type-safe approach instead of returning bad values.

jkbradley commented:
@MechCoder Thanks for the updates!  We'll need to use traits to fix the multiple inheritance issue:
```
sealed trait LogisticRegressionSummary
sealed trait LogisticRegressionTrainingSummary
class BinaryLogisticRegressionSummary extends LogisticRegressionSummary
class BinaryLogisticRegressionTrainingSummary extends BinaryLogisticRegressionSummary with LogisticRegressionTrainingSummary
```
Notes:
* LogisticRegressionSummary could be an abstract class if it saves a lot of code, but if not, I'd prefer it remain a trait.
* I'm marking the traits sealed so that we can extend them in the future without breaking the public API.
* These classes and traits will need to be public.

Let me know if you have alternate ideas for these items.

MechCoder commented:
With regards to the trait, I have a few queries.

Are we planning to have a `MulticlassLogisticRegressionSummary` inheriting from `LogisticRegressionSummary` in the future because without that I'm unable to understand how using a trait would help since there is no access to the predictions dataframe.

MechCoder committed:
Clean up class inheritance

MechCoder commented:
OK, I've fixed up your comment. But can you give a concrete example of how not using a sealed trait will break the API?

Is it because a set of matches cannot be done by the user on the returned Summary object, if the trait is not sealed?

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39675 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39675/consoleFull) for   PR 7538 at commit [`d775371`](https://github.com/apache/spark/commit/d775371deb987d22f1469a3b10b3d97b850615da).

SparkQA commented:
[Test build #39675 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39675/console) for   PR 7538 at commit [`d775371`](https://github.com/apache/spark/commit/d775371deb987d22f1469a3b10b3d97b850615da).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `sealed trait LogisticRegressionTrainingSummary extends Serializable `
  * `sealed trait LogisticRegressionSummary extends Serializable `

AmplabJenkins commented:
Merged build finished. Test FAILed.

MechCoder commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #201 has started](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/201/consoleFull) for   PR 7538 at commit [`d775371`](https://github.com/apache/spark/commit/d775371deb987d22f1469a3b10b3d97b850615da).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39685 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39685/consoleFull) for   PR 7538 at commit [`d775371`](https://github.com/apache/spark/commit/d775371deb987d22f1469a3b10b3d97b850615da).

SparkQA commented:
[Test build #201 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/201/console) for   PR 7538 at commit [`d775371`](https://github.com/apache/spark/commit/d775371deb987d22f1469a3b10b3d97b850615da).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `sealed trait LogisticRegressionTrainingSummary extends Serializable `
  * `sealed trait LogisticRegressionSummary extends Serializable `

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #39685 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39685/console) for   PR 7538 at commit [`d775371`](https://github.com/apache/spark/commit/d775371deb987d22f1469a3b10b3d97b850615da).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `sealed trait LogisticRegressionTrainingSummary extends Serializable `
  * `sealed trait LogisticRegressionSummary extends Serializable `

AmplabJenkins commented:
Merged build finished. Test PASSed.

jkbradley commented:
> Are we planning to have a MulticlassLogisticRegressionSummary inheriting from LogisticRegressionSummary in the future because without that I'm unable to understand how using a trait would help since there is no access to the predictions dataframe.

Yes, MulticlassLogisticRegressionSummary should be analogous to the binary version, with both inheriting from LogisticRegressionSummary.

> can you give a concrete example of how not using a sealed trait will break the API?

Adding a method to a trait is a breaking API change.  If a user has implemented some class which extends the trait, then adding a method to the trait will mean the user's class will no longer implement all of the methods it needs to.  Marking it sealed will prevent users from extending the trait so that we can add more methods in the future.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -286,6 +294,41 @@ class LogisticRegressionModel private[ml] (
 
   override val numClasses: Int = 2
 
+  private var trainingSummary: Option[BinaryLogisticRegressionTrainingSummary] = None
```
should be generic summary, not binary one

MechCoder replied:
Now it makes sense. Sorry about this.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -286,6 +294,41 @@ class LogisticRegressionModel private[ml] (
+   * Gets summary of model on training set. An exception is
+   * thrown if `trainingSummary == None`.
+   */
+  def summary: BinaryLogisticRegressionTrainingSummary = trainingSummary match {
```
This and other methods within LogisticRegressionModel should return LogisticRegressionTrainingSummary or LogisticRegressionSummary, not specific binary/multiclass versions.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -286,6 +294,41 @@ class LogisticRegressionModel private[ml] (
+  }
+
+  private[classification] def setSummary(
+      summary: BinaryLogisticRegressionTrainingSummary): this.type = {
```
should take LogisticRegressionTrainingSummary

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -286,6 +294,41 @@ class LogisticRegressionModel private[ml] (
+  // TODO: decide on a good name before exposing to public API
+  private[classification] def evaluate(dataset: DataFrame): BinaryLogisticRegressionSummary = {
+    val t = udf { features: Vector => raw2probabilityInPlace(predictRaw(features)) }
+    val labelsAndScores = dataset.
```
It'd be better just to call ```this.transform(dataset)``` and pass that to the summary constructor.  Less code, and it keeps more columns.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -408,6 +451,119 @@ private[classification] class MultiClassSummarizer extends Serializable {
 /**
+ * Abstraction for multinomial Logistic Regression Training results.
+ */
+sealed trait LogisticRegressionTrainingSummary extends Serializable {
```
This should extend trait LogisticRegressionSummary, to be precise.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -408,6 +451,119 @@ private[classification] class MultiClassSummarizer extends Serializable {
+sealed trait LogisticRegressionTrainingSummary extends Serializable {
+
+  /** objective function (scaled loss + regularization) at each iteration. */
+  val objectiveHistory: Array[Double]
```
I'd prefer this be marked as a def.  Not sure if it matters, but that's what I've seen done.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -408,6 +451,119 @@ private[classification] class MultiClassSummarizer extends Serializable {
+  val objectiveHistory: Array[Double]
+
+  /** Number of training iterations until termination */
+  val totalIterations: Int = objectiveHistory.length
```
I think this is OK, but we should test this in a little Java suite.  Implementations within traits can cause Java API problems.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -408,6 +451,119 @@ private[classification] class MultiClassSummarizer extends Serializable {
+/**
+ * Abstraction for Logistic Regression Results for a given model.
+ */
+sealed trait LogisticRegressionSummary extends Serializable {
```
This trait can have defs for predictions, probabilityCol, labelCol.

MechCoder replied:
I am surprised this works in spite of these being vals in the subclass.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala:
```diff
@@ -408,6 +451,119 @@ private[classification] class MultiClassSummarizer extends Serializable {
+  /**
+   * Computes the area under the receiver operating characteristic (ROC) curve.
+   */
+  def areaUnderROC(): Double = binaryMetrics.areaUnderROC()
```
Make lazy val

jkbradley commented:
Just a few items remain

MechCoder committed:
Change defs into lazy vals

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #40027 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40027/consoleFull) for   PR 7538 at commit [`2e9f7c7`](https://github.com/apache/spark/commit/2e9f7c724a14d7170e0dd910195bde00bb54bd89).

SparkQA commented:
[Test build #40027 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40027/console) for   PR 7538 at commit [`2e9f7c7`](https://github.com/apache/spark/commit/2e9f7c724a14d7170e0dd910195bde00bb54bd89).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `sealed trait LogisticRegressionTrainingSummary extends LogisticRegressionSummary `
  * `sealed trait LogisticRegressionSummary extends Serializable `

AmplabJenkins commented:
Merged build finished. Test PASSed.

MechCoder commented:
@jkbradley I have addressed your comments in the last commit. I have a few last minor questions.

1. For accessing the `pr()`, `fMeasureByThreshold` etc, I'll have to do model.summary.asInstanceOf[Binary..] I suppose that should be okay, right? (Similar things are being done in LDAModel etc)
2. If I make objectiveHistory and totalIterations defs then, that would be different from the LinearRegressionSummary where it would be vals. This would create differences when being called from Java.

jkbradley commented:
> For accessing the pr(), fMeasureByThreshold etc, I'll have to do model.summary.asInstanceOf[Binary..] I suppose that should be okay, right? (Similar things are being done in LDAModel etc)

I agree it's a bit awkward, but I prefer that to providing null/bad values.  The other big choice we could have made when creating spark.ml is separate binary and multiclass algorithms, but that would have created a bunch of copied APIs.

> If I make objectiveHistory and totalIterations defs then, that would be different from the LinearRegressionSummary where it would be vals. This would create differences when being called from Java. i.e, I'll have to do objectiveHistory() for Logistic nd objectiveHistory for Linear

I don't think def and val look different from Java.  The Scala compiler creates both as methods, so they should appear to be the same for the Java and Scala APIs.

LGTM.  Thanks for iterating through updates with me!  I'll merge this with master and branch-1.5

MechCoder commented:
All right, the reason I thought it would be different is that in the last commit (https://github.com/MechCoder/spark/commit/2e9f7c724a14d7170e0dd910195bde00bb54bd89#diff-1747fe912f0ee426f29b3613e6b0a197R156) just doing `model.summary` raised an error, while in Scala it works.

jkbradley commented:
Oh, I see.  That's because Scala can have method calls without parentheses, whereas Java requires the parentheses.

MechCoder commented:
Yes, that's what I had meant. Would that be okay?

jkbradley commented:
Yes, that's fine since the same method works in both languages.

MechCoder commented:
Should I open a JIRA to refactor again into a general `RegressionSummary` and `ClassificationSummary` since almost all of these metrics would be common to all algorithms? I can open a JIRA for the setBins as well.

jkbradley commented:
Sure, the refactoring sounds great, thanks!  Please link to the R-like stats for models JIRA.
I think setBins is low priority for now but would be good eventually.
